Combine InteractiveMarkerViewer classes.

aditya-vk opened the pull request:
Simplifies the code in InteractiveMarkerViewer. Refer to #520 for details (followed suggestion 2).

Resolves #519. Resolves #520.

***

**Before creating a pull request**

- [x] Document new methods and classes
- [x] Format code with `make format`

**Before merging a pull request**

- [x] Set version target by selecting a milestone on the right side
- [x] Summarize this change in `CHANGELOG.md`
- [ ] Add unit test(s) for this change

Kelym commented on src/rviz/InteractiveMarkerViewer.cpp:
```diff
@@ -168,34 +164,52 @@ void InteractiveMarkerViewer::update()
+      {
+        it = mSkeletonMarkers.erase(it);
+      }
     }
```
To confirm: we don't need to update `mSkeletonMarkers`even if we don't have `mWorld`?

Kelym commented on include/aikido/rviz/InteractiveMarkerViewer.hpp:
```diff
@@ -24,8 +25,14 @@ AIKIDO_DECLARE_POINTERS(InteractiveMarkerViewer)
 class InteractiveMarkerViewer
 {
 public:
+  /// Create an InteractiveMarkerViewer that reflects skeletons in a World.
```
```suggestion
  /// Creates an InteractiveMarkerViewer that reflects skeletons in a World.
```

Kelym commented on include/aikido/rviz/InteractiveMarkerViewer.hpp:
```diff
@@ -24,8 +25,14 @@ AIKIDO_DECLARE_POINTERS(InteractiveMarkerViewer)
+  /// Create an InteractiveMarkerViewer that reflects skeletons in a World.
+  /// \param[in] topicNamespace ROS topic to publish marker updates to
+  /// \param[in] frameId Base frame name
+  /// \param[in] env World to update viewer with
```
```suggestion
  /// \param[in] env World to update the viewer with
```

Kelym commented on include/aikido/rviz/InteractiveMarkerViewer.hpp:
```diff
@@ -34,22 +41,29 @@ class InteractiveMarkerViewer
+
+  /// Visualizes a Frame as a cylinder.
+  /// \param[in] frame Target DART frame.
+  /// \param[in] length Length of the cylinderical frame (along z-axis).
```
```suggestion
  /// \param[in] length Length of the cylindrical frame (along z-axis).
```

Kelym commented on include/aikido/rviz/InteractiveMarkerViewer.hpp:
```diff
@@ -34,22 +41,29 @@ class InteractiveMarkerViewer
       double thickness = 0.02,
       double alpha = 1.0);
 
-  SkeletonMarkerPtr CreateSkeletonMarker(
```
It looks like a handy helper for creating `SkeletonMarker`. You can use it in `addSkeleton` and `update`. Why bother removing it?

aditya-vk replied:
I preferred consistency (across skeletons, frames etc.) over code duplication. We could also have `createFrameMarker`, `createTSRMarker` etc. I just removed it instead.

Kelym commented on include/aikido/rviz/InteractiveMarkerViewer.hpp:
```diff
@@ -77,28 +91,51 @@ class InteractiveMarkerViewer
       double thickness = 0.01,
       std::size_t numLineSegments = 16u);
 
+  /// Set viewer auto-updating.
```
```suggestion
  /// Sets viewer auto-updating to on (true) or off.
```

Kelym commented on include/aikido/rviz/InteractiveMarkerViewer.hpp:
```diff
@@ -34,22 +41,29 @@ class InteractiveMarkerViewer
+  /// Visualizes a Skeleton.
+  /// \param skeleton Skeleton to add to the viewer
+  /// \return Skeleton marker added to the viewer.
+  SkeletonMarkerPtr addSkeletonMarker(
```
So if the Viewer is initialized without `mWorld`, can one still call this function? If not, can we update the docstring and set a check inside the function.

aditya-vk replied:
Yes.

Kelym commented on include/aikido/rviz/InteractiveMarkerViewer.hpp:
```diff
@@ -77,28 +91,51 @@ class InteractiveMarkerViewer
+  /// \param[in] flag Whether to auto-update the viewer.
   void setAutoUpdate(bool flag);
+
+  /// Update viewer with Skeletons from the World and existing markers.
```
```suggestion
  /// Updates viewer with Skeletons from the World and existing markers.
```

Kelym commented on include/aikido/rviz/InteractiveMarkerViewer.hpp:
```diff
@@ -77,28 +91,51 @@ class InteractiveMarkerViewer
   interactive_markers::InteractiveMarkerServer mMarkerServer;
-  std::set<SkeletonMarkerPtr> mSkeletonMarkers;
+
+  /// Mapping of Skeletons to SkeletonMarkers
```
```suggestion
  /// Map of Skeletons to SkeletonMarkers
```

aditya-vk commented:
@Kelym @brianhou Addressed the comments. Added a docstring that `mWorld` can be a `nullptr` if we are just adding markers, and otherwise the skeletons in the world shall also get updated. Could you review?

Kelym commented on src/rviz/InteractiveMarkerViewer.cpp:
```diff
@@ -168,34 +164,60 @@ void InteractiveMarkerViewer::update()
-        (*it)->update();
-
-      ++it;
+        it->second->update();
```
```suggestion
        it->second->update();
        ++it;
```

aditya-vk commented:
@Kelym I updated the formatting and the ChangeLog. Should be ready for merging into master?

brianhou commented on CHANGELOG.md:
```diff
@@ -82,6 +82,7 @@
 
   * Fixed bug of not joining Viewer threads when stopping auto-update: [#463](https://github.com/personalrobotics/aikido/pull/463)
   * Fixed bug of not passing full file path to RViz when MeshShape is used: [#518](https://github.com/personalrobotics/aikido/pull/518)
+  * Merge WorldInteractiveMarkerViewer and InteractiveMarkerViewer classes, removing the former: [#520](https://github.com/personalrobotics/aikido/pull/520)
```
```suggestion
  * Merged WorldInteractiveMarkerViewer into InteractiveMarkerViewer, removing the former: [#520](https://github.com/personalrobotics/aikido/pull/520)
```
